Don't export EbmlMaster context, use the class static member

robUx4 opened the pull request:
The internal are reworked so only semantic context of EbmlMaster elements have a table of possible children.

"[Get]ElementSpec()" is generalized so there's stronger typing of the semantic context for all the types.

The callback to get the global elements now returns a `EbmlSemanticContextMaster`. Ultimately it only needs to return a container with a list of semantic context.

mbunkus commented:
Unfortunately this breaks [this function](https://gitlab.com/mbunkus/mkvtoolnix/-/blob/main/src/common/ebml.cpp#L79) & several other similar ones in MKVToolNix. The error message is:

```
src/common/ebml.cpp:88:15: fatal error: no matching function for call to 'tEBML_CTX_IDX'
    if (id == EBML_CTX_IDX_ID(context,i))
              ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:274:45: note: expanded from macro 'EBML_CTX_IDX_ID'
#define EBML_CTX_IDX_ID(c,i)   EBML_INFO_ID(EBML_CTX_IDX_INFO(c,i))
                                            ^~~~~~~~~~~~~~~~~~~~~~
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:273:47: note: expanded from macro 'EBML_CTX_IDX_INFO'
#define EBML_CTX_IDX_INFO(c,i) EBML_SEM_SPECS(EBML_CTX_IDX(c,i))
                                              ^~~~~~~~~~~~~~~~~
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:272:32: note: expanded from macro 'EBML_CTX_IDX'
#define EBML_CTX_IDX(c,i)      tEBML_CTX_IDX(c,i)
                               ^~~~~~~~~~~~~
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:265:45: note: expanded from macro 'EBML_SEM_SPECS'
#define EBML_SEM_SPECS(s)   tEBML_SEM_SPECS(s)
                                            ^
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:260:45: note: expanded from macro 'EBML_INFO_ID'
#define EBML_INFO_ID(cb)      tEBML_INFO_ID(cb)
                                            ^~
/home/mosu/prog/video/mkvtoolnix/lib/libebml/ebml/EbmlElement.h:531:36: note: candidate function not viable: no known conversion from 'const EbmlSemanticContext' to 'const EbmlSemanticContextMaster' for 1st argument
static inline const EbmlSemantic & tEBML_CTX_IDX(const EbmlSemanticContextMaster & c, std::size_t i)
                                   ^
1 error generated.
```

Any pointers how I would have to change my code? I need to iterate over the whole semantic tree & find elements, either by their ID, or by their debug name.

mbunkus commented:
Compilation still fails after merging 9e6f5b5b1d47b9e0fb275ad5e2d7f2a408da7327 (& 'hide_context' for libMatroska) with the same error message I've posted above. Meaning the same issue remains; I need to be able to iterate over the whole structure, and I don't know how after these changes.

robUx4 commented:
OK, I see the problem. Now that there's stronger typing you can only get a child `EbmlSemantic` from an `EbmlSemanticContextMaster`. You doesn't make much sense for an integer or a binary context. (I'd like to split this even more, but without relying on RTTI).

I'll make the patches for mkvtoolnix.

mbunkus commented:
As written over on Gitlab, but to keep the conversation here: I needed a couple of fixes on top of your MKVToolNix PR to make MKVToolNix compile. Afterwards a handful of my test cases fail.

I've now found out why they fail: `SkipData()` works differently for elements with unknown sizes after merging this PR. Before the PR calling `SkipData()` on an element (a cluster in this case) with an unknown size caused the file pointer to be set to the end of the file, causing further reads to fail. That was fine for me.

After the PR the file pointer is now directly at the start of the cluster's data portion (directly after the cluster's head).

I consider the new behavior to be bad & wrong.

A typical algorithm used by MKVToolNix is to create a list of level 1 elements like this:

1. read the EBML head
2. find the segment
3. while not at end of file:
  1. find next ID
  2. read element head
  3. make note of its position & size
  4. skip over it

This worked fine with the old behavior. With the new one the "skip over it" part will leave the file pointer after the level 1 element's head, meaning that the next "find next ID" call will find the ID of the first child element of that level 1 element.

That is a substantial change in semantics. Sure, I could work around this, but I shouldn't have to, in my opinion, as the new behavior is strictly worse to the old one.

Unfortunately I really don't understand why the changes in this PR affect skipping data that way & can therefore not offer a fix myself.

mbunkus commented:
BTW, the file in question is [this one](https://mkvtoolnix.download/misc/live-stream.webm).

robUx4 commented:
I removed the `FindChildByEbmlId()` commits for now. That's new code to simplify getting a child semantic. It's not needed in this PR and may be the source of the regression. I also updated https://github.com/Matroska-Org/libmatroska/pull/183 to match this PR.

robUx4 commented:
I think I identified the source of the issue. The `if (EltIndex >= EBML_CTX_SIZE(Context)) {` par of the code was no longer executed in `SkipData()` when a child element was no found for infinite size elements.
